### ISSUE TITLE
undefined method `get_app_info_localizations' for nil

### ISSUE THREAD

**Issue by MarMarIV** [fastlane/fastlane#17208]
I get fastlane error, when using this command:

fastlane precheck -u mail@domain.com -a app_ident

Any suggestions to solve problem?

**Log:**
INFO [2020-09-10 13:05:18.47]: Successfully logged in
INFO [2020-09-10 13:05:18.47]: Checking app for precheck rule violations

Looking for related GitHub issues on fastlane/fastlane...
Search query: undefined method `get_app_info_localizations' for nil

URL: https://api.github.com/search/issues?q=undefined%20method%20%60get_app_info_localizations%27%20for%20nil+repo:fastlane/fastlane
Found no similar issues. To create a new issue, please visit:
https://github.com/fastlane/fastlane/issues/new
Run `fastlane env` to append the fastlane environment to your issue
Traceback (most recent call last):
	15: from /usr/local/bin/fastlane:23:in `<main>'
	14: from /usr/local/bin/fastlane:23:in `load'
	13: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/bin/fastlane:23:in `<top (required)>'
	12: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/fastlane/lib/fastlane/cli_tools_distributor.rb:111:in `take_off'
	11: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/commands_generator.rb:19:in `start'
	10: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/commands_generator.rb:71:in `run'
	 9: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	 8: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	 7: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	 6: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	 5: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	 4: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/commands_generator.rb:41:in `block (2 levels) in run'
	 3: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/runner.rb:34:in `run'
	 2: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/runner.rb:102:in `check_for_rule_violations'
	 1: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/rule_processor.rb:47:in `process_app_and_version'
/Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/rule_processor.rb:134:in `generate_app_items_to_check': undefined method `get_app_info_localizations' for nil:NilClass (NoMethodError)
	15: from /usr/local/bin/fastlane:23:in `<main>'
	14: from /usr/local/bin/fastlane:23:in `load'
	13: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/bin/fastlane:23:in `<top (required)>'
	12: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/fastlane/lib/fastlane/cli_tools_distributor.rb:111:in `take_off'
	11: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/commands_generator.rb:19:in `start'
	10: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/commands_generator.rb:71:in `run'
	 9: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	 8: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	 7: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	 6: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	 5: from /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	 4: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/commands_generator.rb:41:in `block (2 levels) in run'
	 3: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/runner.rb:34:in `run'
	 2: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/runner.rb:102:in `check_for_rule_violations'
	 1: from /Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/rule_processor.rb:47:in `process_app_and_version'
/Library/Ruby/Gems/2.6.0/gems/fastlane-2.158.0/precheck/lib/precheck/rule_processor.rb:134:in `generate_app_items_to_check': [!] undefined method `get_app_info_localizations' for nil:NilClass (NoMethodError)

**Env:**
<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.15.6                                                        |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.24.3 (Apple Git-128)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | Mac OS X 10.15.6 (19G2021)                                     |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 11.7                                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.158.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.3        |
| CFPropertyList            | 2.3.6        |
| claide                    | 1.0.3        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.3.0        |
| xcodeproj                 | 1.18.0       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.7.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| public_suffix             | 4.0.5        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.8.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| excon                     | 0.76.0       |
| faraday                   | 1.0.1        |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 1.0.0        |
| fastimage                 | 2.2.0        |
| gh_inspector              | 1.1.3        |
| json                      | 2.1.0        |
| mini_magick               | 4.10.1       |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.7.6        |
| bundler                   | 1.17.2       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| jwt                       | 2.2.2        |
| uber                      | 0.1.0        |
| declarative               | 0.0.20       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| multi_json                | 1.15.0       |
| signet                    | 0.14.0       |
| memoist                   | 0.16.2       |
| os                        | 1.1.1        |
| googleauth                | 0.13.1       |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| google-cloud-env          | 1.3.3        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-core         | 1.5.0        |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-storage      | 1.27.0       |
| emoji_regex               | 3.0.0        |
| jmespath                  | 1.4.0        |
| aws-partitions            | 1.358.0      |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.2        |
| aws-sdk-core              | 3.104.4      |
| aws-sdk-kms               | 1.36.0       |
| aws-sdk-s3                | 1.78.0       |
| forwardable               | 1.2.0        |
| logger                    | 1.3.0        |
| stringio                  | 0.0.2        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.0        |
| date                      | 2.0.0        |
| fileutils                 | 1.1.0        |
| etc                       | 1.0.1        |
| io-console                | 0.4.7        |
| zlib                      | 1.0.0        |
| libxml-ruby               | 3.1.0        |
| rexml                     | 3.1.9        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.4.2        |
</details>

*generated on:* **2020-09-10**
</details>

**Comment by jopicornell** [comment]
I'm facing the same problem. Any Idea why is this happening? Seems like Apple has changed some endpoint making the endpoint for localizations fail

**Comment by khallmark** [comment]
Also getting this error.

**Comment by max-ott** [comment]
I don't think this is yet using the new `ConnectAPI` and therefore fails. More than happy to look into any PR.

**Comment by cambrantskv** [comment]
> I don't think this is yet using the new `ConnectAPI` and therefore fails. More than happy to look into any PR.

Thank you for the information and advice!

**Comment by jcharr1** [comment]
We're suddenly getting this for the last couple of days or so, too. I tried downgrading fastlane to versions 2.161.0 through 2.163.0 but it didn't make a difference. Maybe something changed or there's some kind of issue on Apple's side?

**Comment by jcharr1** [comment]
Never mind. Whatever the issue on Apple's end was appears to be resolved now.

**Comment by nero2009** [comment]
Still getting this error

**Comment by jcharr1** [comment]
After my last post it's been intermittently happening to us too.

**Comment by jcharr1** [comment]
It seems to no longer be intermittent anymore, and now happens all the time since my last post.

**Comment by leo-pedidosya** [comment]
Still getting this error

**Comment by rogerluan** [comment]
I could reproduce this issue consistently when my app was live and had no edit version available. Can you guys confirm whether you actually have an editable app before calling precheck? 

States that we consider to be editable:

```ruby
states = [
  Spaceship::ConnectAPI::AppInfo::AppStoreState::PREPARE_FOR_SUBMISSION,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::DEVELOPER_REJECTED,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::REJECTED,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::METADATA_REJECTED,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::WAITING_FOR_REVIEW,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::INVALID_BINARY
]
```

States that are not editable:

```ruby
states = [
  Spaceship::ConnectAPI::AppInfo::AppStoreState::READY_FOR_SALE,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::PROCESSING_FOR_APP_STORE,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::PENDING_DEVELOPER_RELEASE,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::PENDING_APPLE_RELEASE,
  Spaceship::ConnectAPI::AppInfo::AppStoreState::IN_REVIEW
]
```

If you guys can validate my assumption, we should improve this by avoiding the crash and throwing a nice error message instead.

**Comment by jacque006** [comment]
@rogerluan We are encountering this issue on our release for iOS. We haven't modified our `Fastfile` in quite some time so it's perplexing why this is happening now.

Relevant part of our iOS `Fastfile`:
```ruby
platform :ios do

  ...

  private_lane :compile_app do |channel: "appstore", configuration: "Release", **options|
    # Set up some vars
    signing_type = "appstore"
    app_name = APP_NAME
    app_identifier = APP_IDENTIFIER
    version_number = ENV["REDACTED"]
    build_number = ENV["REDACTED"] || raise("Missing REDACTED")
    increment_build_number xcodeproj: IOS_PROJECT, build_number: build_number

    # If a channel is specfied, update the project to reflect it
    set_envfile(channel: channel)
    scheme = "REDACTED #{channel}"

    if channel != "appstore"
      app_name = app_name + " (#{channel})"
      app_identifier = "REDACTED.#{channel}"
      version_number = get_version_number xcodeproj: IOS_PROJECT, target: "REDACTED"
      label_icons top: channel, bottom: "v#{version_number}b#{build_number}"
    # appstore requires no 'v' in version_number
    else
      version_number = version_number.sub "v", ""
    end

    increment_version_number(
      xcodeproj: IOS_PROJECT,
      version_number: version_number
    )

    update_info_plist(
      app_identifier: app_identifier,
      xcodeproj: IOS_PROJECT,
      plist_path: "REDACTED/Info.plist",
    )

    update_info_plist(
      xcodeproj: IOS_PROJECT,
      plist_path: "REDACTED/Info.plist",
      block: lambda { |plist|
          urlScheme = plist["CFBundleURLTypes"].find{|scheme| scheme["CFBundleURLSchemes"] == ["REDACTED"]}
          urlScheme[:CFBundleURLSchemes] = [ENV["SCHEME"]]
          plist["REDACTED"] = ENV["REDACTED"]
      }
    )

    update_plist(
      plist_path: "REDACTED.entitlements.release",
      block: lambda { |plist|
          plist["com.apple.developer.associated-domains"] = [
              "applinks:#{ENV["REDACTED"]}",
          ]
      }
    )

    # Ensure the app is created on iTunes Connect
    produce(
      app_identifier: app_identifier,
      app_name: app_name,
      language: "English",
      app_version: version_number,
    )

    # Sync the code signing
    sync_code_signing(
      app_identifier: app_identifier,
      type: signing_type,
      git_url: "REDACTED"
    )

    # Update the provisioning profile specifier
    update_project_provisioning(
      xcodeproj: IOS_PROJECT,
      profile: ENV["sigh_#{app_identifier}_appstore_profile-path"],
      build_configuration: configuration
    )

    # Build the app
    build_ios_app(
      **options,
      scheme: scheme,
      configuration: configuration,
      codesigning_identity: "REDACTED",
      export_team_id: "REDACTED",
      workspace: IOS_WORKSPACE,
      silent: true,
    )

    # Return the app id
    app_identifier
  end

...

  desc "Release a version to the app store"
  lane :release do
    raise "Can only release from REDACTED" unless (version_number = ENV["REDACTED"])
    create_release_notes
    check_app_store_metadata # <- fails with this error here
    app_identifier = compile_app(channel: "appstore")
    upload_to_app_store(
      app_identifier: app_identifier,
      force: true,
      ignore_language_directory_validation: true,
      submit_for_review: true,
      automatic_release: true,
    )
  end
end
```
I was able to manually release by commenting out the call to `check_app_store_metadata`.
Did something change recently where we need to move `check_app_store_metadata` to a later part of our release process when our `AppStoreState` will be in one of the editable states you listed above?

**Comment by jcharr1** [comment]
An interesting thing I've noticed that when we run precheck manually after uploading a build to TestFlight it fails with this error. But when it's run when we submit the build for review and it runs precheck as part of deliver, it works.

I typically don't create a new version on App Store Connect until we're ready to submit very soon. @rogerluan I just tried running it with no pending version created and it failed. After I created a new version on App Store Connect it worked.

**Comment by rogerluan** [comment]
Thanks for the input, everyone.

The comments posted after my last message here confirmed my assumptions, where only "editable" app versions were able to be prechecked. However, based on the testimonials, this sounded like a regression. So I investigated further [in _precheck_ history](https://github.com/fastlane/fastlane/commits/master/precheck/lib/precheck/rule_processor.rb), and looks like prior to Sep 4th 2020 _precheck_ could be used to check metadata of both edit **and** live app versions.

As I don't see any reason to limit it to only editable versions, I modified the code to allow live versions to be prechecked as well.

Here's the PR: https://github.com/fastlane/fastlane/pull/17738

If you guys could help us test it, that'd be super helpful! 🎉  @MarMarIV @jacque006 @jcharr1 

Thanks everyone who provided more info so this could be resolved ❤️
